librz/asm/arch/sh/disassembler.c

```diff
@@ -0,0 +1,223 @@
+// SPDX-FileCopyrightText: 2022 Dhruv Maroo <dhruvmaru007@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include "disassembler.h"
+#include "common.h"
+#include "regs.h"
+
+/**
+ * \brief Get SHParam from opcode
+ * Make sure the opcode is passed in little-endian form
+ *
+ * \param opcode instruction opcode (assumed to be in little-endian)
+ * \param shb SHParamBuilder instance which contains the necessary info to find the param
+ */
+static SHParam sh_op_get_param(ut16 opcode, SHParamBuilder shb) {
+	if (shb.is_param) {
+		return shb.param;
+	}
+	if (shb.addr.mode == SH_ADDR_INVALID) {
+		SHParam invalid;
+		invalid.mode = SH_ADDR_INVALID;
+		return invalid;
+	}
+
+	ut16 nibble = opcode >> shb.addr.start;
+	ut8 len = 0;
+
+	switch (shb.addr.mode) {
+	case SH_REG_DIRECT:
+	case SH_REG_INDIRECT:
+	case SH_REG_INDIRECT_I:
+	case SH_REG_INDIRECT_D:
+	case SH_REG_INDIRECT_INDEXED:
+	case SH_PC_RELATIVE_REG:
+		len = 4;
+		break;
+	case SH_REG_INDIRECT_DISP:
+	case SH_GBR_INDIRECT_DISP:
+	case SH_PC_RELATIVE_DISP:
+	case SH_PC_RELATIVE8:
+	case SH_IMM_S:
+	case SH_IMM_U:
+		len = 8;
+		break;
+	case SH_PC_RELATIVE12:
+		len = 12;
+		break;
+	default:
+		break;
+	}
+
+	nibble &= 0xffff >> (16 - len);
+
+	SHParam ret_param;
+	ret_param.mode = shb.addr.mode;
+	switch (shb.addr.mode) {
+	case SH_REG_DIRECT:
+	case SH_REG_INDIRECT:
+	case SH_REG_INDIRECT_I:
+	case SH_REG_INDIRECT_D:
+	case SH_REG_INDIRECT_INDEXED:
+	case SH_GBR_INDIRECT_DISP:
+	case SH_PC_RELATIVE_DISP:
+	case SH_PC_RELATIVE8:
+	case SH_PC_RELATIVE12:
+	case SH_PC_RELATIVE_REG:
+	case SH_IMM_S:
+	case SH_IMM_U:
+		ret_param.param[0] = nibble;
+		break;
+	case SH_REG_INDIRECT_DISP:
+		ret_param.param[0] = nibble >> 4;
+		ret_param.param[1] = nibble & 0xf;
+		break;
+	case SH_GBR_INDIRECT_INDEXED:
+		break;
+	default:
+		break;
+	}
+
+	return ret_param;
+}
+
+/**
+ * \brief Get params for mov.l instruction (0001NMD)
+ * A special function is required because the nibbles for the second param (@(disp:Rn)) (i.e. N and D)
+ * are separated by the nibble for the first param (Rm) (i.e. M), so sh_op_get_param cannot be used
+ *
+ * \param opcode opcode
+ * \param m if true, get Rm ; otherwise get @(disp:Rn)
+ * \return SHParam return appropriate param
+ */
+static SHParam sh_op_get_param_movl(ut16 opcode, bool m) {
+	if (m) {
+		ut16 reg = (opcode >> 4) & 0xf;
+		return (SHParam){ .param = {
+					  reg,
+				  },
+			.mode = SH_REG_DIRECT };
+	} else {
+		ut16 d = opcode & 0xf;
+		ut16 n = (opcode >> 8) & 0xf;
+		return (SHParam){ .param = { n, d }, .mode = SH_REG_INDIRECT_DISP };
+	}
+}
+
+/**
+ * \brief Disassemble \p opcode and return a SHOp
+ *
+ * \param opcode 16 bit wide opcode
+ * \return SHOp object corresponding to the opcode
+ */
+RZ_API RZ_OWN SHOp *sh_disassembler(ut16 opcode) {
+	for (ut16 i = 0; i < OPCODE_NUM; i++) {
+		if ((opcode | sh_op_lookup[i].mask) == sh_op_lookup[i].opcode) {
```

Comment:
You could reverse the condition and call `continue` instead, one level of nestiness would be gone.